File: src/storage/lua_scripts/leaky_bucket.lua
local key = KEYS[1]
local now = tonumber(ARGV[1])
local capacity = tonumber(ARGV[2])
local leak_rate = tonumber(ARGV[3]) -- tokens/sec drained
local requested = tonumber(ARGV[4]) or 1

-- Retrieve existing state
local data = redis.call('HMGET', key, 'water_level', 'last_leak_time')
local water_level = tonumber(data[1])
local last_leak_time = tonumber(data[2])

if not water_level or not last_leak_time then
    -- First request, initialize bucket
    water_level = requested
    last_leak_time = now
    redis.call('HMSET', key, 'water_level', water_level, 'last_leak_time', last_leak_time)
    local ttl = math.ceil(capacity / leak_rate)
    redis.call('EXPIRE', key, ttl)
    return {1, capacity - water_level}
else
    -- Leak water based on elapsed time
    local elapsed = math.max(0, now - last_leak_time)
    local leaked = elapsed * leak_rate
    water_level = math.max(0.0, water_level - leaked)
    last_leak_time = now
    
    if water_level + requested <= capacity then
        water_level = water_level + requested
        redis.call('HMSET', key, 'water_level', water_level, 'last_leak_time', last_leak_time)
        local ttl = math.ceil(capacity / leak_rate)
        redis.call('EXPIRE', key, ttl)
        return {1, capacity - water_level}
    else
        -- Save the leaked state anyway
        redis.call('HMSET', key, 'water_level', water_level, 'last_leak_time', last_leak_time)
        local ttl = math.ceil(capacity / leak_rate)
        redis.call('EXPIRE', key, ttl)
        return {0, capacity - water_level}
    end
end

File: src/enforcement/algorithms/leaky_bucket.py
import time
import logging
from typing import Tuple, Dict
from src.enforcement.algorithms.base import BaseRateLimiter
from src.storage.redis_client import RedisClient

logger = logging.getLogger("arl.enforcement.leaky_bucket")

class LeakyBucketLimiter(BaseRateLimiter):
    def __init__(self, redis_client: RedisClient):
        super().__init__(redis_client)
        # In-memory fallback: client_id -> (water_level, last_leak_time)
        self._fallback_db: Dict[str, Tuple[float, float]] = {}

    async def check_limit(self, client_id: str, rate: float, capacity: float, window_seconds: int, requested: int = 1) -> Tuple[bool, float, float]:
        now = time.time()
        
        # Check if Redis is initialized and connected
        if self.redis_client.client:
            try:
                key = f"rate:lb:{client_id}"
                
                # Run Redis Lua script
                # KEYS[1] = key
                # ARGV[1] = now, ARGV[2] = capacity, ARGV[3] = rate (leak_rate), ARGV[4] = requested
                result = await self.redis_client.run_script(
                    "leaky_bucket",
                    keys=[key],
                    args=[str(now), str(capacity), str(rate), str(requested)]
                )
                
                if result is not None:
                    allowed, remaining_capacity = result
                    return bool(allowed), float(remaining_capacity), 0.0
            except Exception as e:
                logger.error(f"Redis Leaky Bucket failed, falling back to in-memory: {e}")

        # Fallback to local in-memory Leaky Bucket
        return self._check_in_memory(client_id, rate, capacity, requested, now)

    def _check_in_memory(self, client_id: str, rate: float, capacity: float, requested: int, now: float) -> Tuple[bool, float, float]:
        state = self._fallback_db.get(client_id)
        if state is None:
            water_level = float(requested)
            self._fallback_db[client_id] = (water_level, now)
            return True, capacity - water_level, 0.0
            
        last_water, last_leak_time = state
        elapsed = max(0.0, now - last_leak_time)
        leaked = elapsed * rate
        water_level = max(0.0, last_water - leaked)
        
        if water_level + requested <= capacity:
            water_level += requested
            self._fallback_db[client_id] = (water_level, now)
            return True, capacity - water_level, 0.0
        else:
            self._fallback_db[client_id] = (water_level, now)
            return False, capacity - water_level, 0.0
            
    def clear_fallback_db(self):
        self._fallback_db.clear()
